### DISCUSSION TITLE
Routing pio interrupts to a specific core

### DISCUSSION THREAD

**Discussion by ardlc-1212** [earlephilhower/arduino-pico#1803] [Q&A] [answered]
Hi I am currently working on a project that includes can2040 library which uses pio to emulate a can controller. it also uses interrupts for things like bit unstuffing and data notification. this library isnt re-entrant or multithread safe so I need to set the interrupts to work on the same core as transmit functions etc.

I couldn't find any documentation 
`
irq_set_exclusive_handler(PIO0_IRQ_0_IRQn, PIOx_IRQHandler);
NVIC_SetPriority(PIO0_IRQ_0_IRQn, 1);
NVIC_EnableIRQ(PIO0_IRQ_0_IRQn);
`
this is the code piece that I use to start and set priority on the interrupt as the can2040 documentation stated. but I don't know which core will enter the routine. is there a way to specify that or is it predetermined to the core that initiates this code piece

Best Regards,
        Arda

**Comment by laser-vectrix** [comment]
Did you have any luck on this? I'm also using can2040 on one core and a wifi (lwip_threadsafe_background) on the 2nd core. At 250kbit/s I get quite some CAN parse errors when TCP connections are running, I guess that is due to PIO interrupts being handled on the 2nd core.

**Reply by ardlc-1212** [reply]
Unfortunately, No.

Due to time constraints I went with a more primitive UART like approach but I will update if I ever get back on the project. Please do so too, if you also discover any issues related.

Best regards,
    Arda

**Comment by earlephilhower** [comment, marked as answer]
Each core has its own interrupt controller, so essentially the core which you call the NVIC commands will be the core that gets the interrupts.  Just make sure only one core is responding to the PIO interrupt and make sure you're running on the right one to do the work...
